tsfresh/feature_extraction/extraction.py

```diff
@@ -103,9 +114,103 @@ def extract_features(timeseries_container, feature_extraction_settings=None,
+
+    :return: The (maybe imputed) DataFrame containing extracted features.
+    :rtype: pandas.DataFrame
+    """
+    all_possible_unique_id_values = set(id_value for kind, df in kind_to_df_map.items()
+                                        for id_value in df[column_id])
+    df_with_ids = pd.DataFrame(index=all_possible_unique_id_values)
+
+    partial_extract_features_for_one_time_series = partial(_extract_features_for_one_time_series,
+                                                           column_id=column_id,
+                                                           column_value=column_value,
+                                                           settings=settings)
+    pool = Pool(settings.n_processes)
+
+    extracted_features = pool.map(partial_extract_features_for_one_time_series, kind_to_df_map.items(),
+                                  chunksize=settings.chunksize)
+
     pool.close()
+
+    # Concatenate all partial results
+    result = pd.concat([df_with_ids] + extracted_features, axis=1, join='outer', join_axes=[df_with_ids.index]) \
+        .astype(np.float64)
+
     pool.join()
+    return result
+
+
+def _extract_features_parallel_per_sample(kind_to_df_map, settings, column_id, column_value):
+    """
+    Parallelize the feature extraction per kind and per sample.
+
+    :param kind_to_df_map: The time series to compute the features for in our internal format
+    :type kind_to_df_map: dict of pandas.DataFrame
+
+    :param column_id: The name of the id column to group by.
+    :type column_id: str
+    :param column_value: The name for the column keeping the value itself.
+    :type column_value: str
+
+    :param settings: settings object that controls which features are calculated
+    :type settings: tsfresh.feature_extraction.settings.FeatureExtractionSettings
 
+    :return: The (maybe imputed) DataFrame containing extracted features.
+    :rtype: pandas.DataFrame
+    """
+    all_possible_unique_id_values = set(id_value for kind, df in kind_to_df_map.items()
+                                        for id_value in df[column_id])
+    df_with_ids = pd.DataFrame(index=all_possible_unique_id_values)
+
+    partial_extract_features_for_one_time_series = partial(_extract_features_for_one_time_series,
+                                                           column_id=column_id,
+                                                           column_value=column_value,
+                                                           settings=settings)
+    pool = Pool(settings.n_processes)
+
+    # Submit map jobs per kind per sample
+    results_fifo = Queue()
+    for kind, df_kind in kind_to_df_map.items():
+        gb = df_kind.groupby(column_id)
+        results_fifo.put(
+            pool.map_async(
+                partial_extract_features_for_one_time_series,
+                [(kind, group) for _, group in gb],
+                chunksize=settings.chunksize
+            )
+        )
+
```

Comment:
1. My first draft did not use an async map. However, splitting the dataframe is quite costly. The idea here is to interleave the splitting, calculation and combining. This results in an overall speed improvement. I should add a comment explaining that.
2. Yes it reads awful. However I'm not sure how to really improve it. I'll try to come up with better  naming of the variables. But it seems to me that for this part you just need to understand the relevant pandas fu.